src/nvim/plines.c

```diff
@@ -991,16 +994,17 @@ int64_t win_text_height(win_T *const wp, const linenr_T start_lnum, const int64_
     lnum = lnum_next + 1;
   }
 
-  while (lnum <= end_lnum) {
+  int64_t max_height = max != NULL ? *max : INT64_MAX;
+  while (lnum <= end_lnum && height_sum_nofill + height_sum_fill < max_height) {
     linenr_T lnum_next = lnum;
     hasFolding(wp, lnum, &lnum, &lnum_next);
     height_sum_fill += win_get_fill(wp, lnum);
     height_cur_nofill = plines_win_nofill(wp, lnum, false);
     height_sum_nofill += height_cur_nofill;
+    lnum_prev = lnum;
```

Comment:
This will use the start of a multi-line fold, but it should be the end instead. I think it should be this:
```suggestion
    lnum_prev = lnum_next;
```
The `lnum_next` variable may a bit misnamed here, but I can't decided how else to name it, as if I name it `lnum_end` it's too similar to `end_lnum`.

Reply:
Oh, I may have named it after the `nextp` argument of `plines_win_full()`.

Reply:
Well technically the last line of a fold doesn't take up any height. It may be more useful to return the last line of a fold but the first line seems less surprising to me.

Reply:
That'll depend on the use case, but I can't think of a concrete use case that actually uses the `end_row` return value. The current use case in LSP floating preview doesn't use that value.

Reply:
I intend to use it in #27855 to determine the last line in a message window (could also use `line('w$')` for that presumably but currently the topline isn't yet what it will be when I want to calculate this...).